qiskit/qpy/objects/schedules.py

```diff
@@ -0,0 +1,435 @@
+        else:
+            raise TypeError(f"Invalid instruction operand type {type_key} for value {data}.")
+        operands.append(value)
+
+    if hasattr(instructions, instruction_class_name):
+        instruction_type = getattr(instructions, instruction_class_name)
+        if label:
+            return instruction_type(*operands, name=label)
+        return instruction_type(*operands)
+
+    raise TypeError(f"Invalid instruction class name {instruction_class_name}.")
+
+
+def _read_alignment_context(file_obj):
+    name_size_raw = file_obj.read(struct.calcsize("!H"))
+    name_size = struct.unpack("!H", name_size_raw)[0]
+    alignment_class_name = file_obj.read(name_size).decode("utf8")
+    kwargs = read_mapping(file_obj)
+
+    if hasattr(alignments, alignment_class_name):
+        alignment_type = getattr(alignments, alignment_class_name)
+        return alignment_type(**kwargs)
+
+    raise TypeError(f"Invalid alignment context name {alignment_class_name}.")
+
+
+def _write_channel(file_obj, data):
+    channel_class_name = data.__class__.__name__.encode("utf8")
+    index_type_key = assign_key(data.index)
+    index_binany = dumps_parameter_value(index_type_key, data.index)
+
+    channel_header = struct.pack(
+        CHANNEL_PACK,
+        len(channel_class_name),
+        index_type_key.encode("utf8"),
+        len(index_binany),
+    )
+    file_obj.write(channel_header)
+    file_obj.write(channel_class_name)
+    file_obj.write(index_binany)
+
+
+def _write_waveform(file_obj, data):
+    samples_binary = dumps_numbers(TypeKey.NUMPY, data.samples)
+    label = data.name
+    if label:
+        label_binary = label.encode("utf8")
+    else:
+        label_binary = bytes()
+
+    wavefom_header = struct.pack(
+        WAVEFORM_PACK,
+        len(label_binary),
+        data.epsilon,
+        len(samples_binary),
+        data.limit_amplitude,
+    )
+    file_obj.write(wavefom_header)
+    file_obj.write(label_binary)
+    file_obj.write(samples_binary)
+
+
+def _write_parametric_pulse(file_obj, data):
+    pulse_class_name = data.__class__.__name__.encode("utf8")
+
+    if hasattr(library, data.__class__.__name__):
+        module_path = bytes()
+    else:
+        # user defined parametric pulse
+        module_path = data.__class__.__module__.encode("utf8")
+
+    if data.name:
+        label = data.name.encode("utf8")
+    else:
+        label = bytes()
+
+    param_pulse_header = struct.pack(
+        PARAMETRIC_PULSE_PACK,
+        len(pulse_class_name),
+        len(module_path),
+        len(label),
+        data.limit_amplitude,
+    )
+    file_obj.write(param_pulse_header)
+    file_obj.write(pulse_class_name)
+    file_obj.write(module_path)
+    file_obj.write(label)
+    write_mapping(file_obj, data.parameters)
+
+
+def _write_measure_processor(file_obj, data):
+    name = data.name.encode("utf8")
+    params = json.dumps(data.params, separators=(",", ":")).encode("utf8")
```

Comment:
I assume parameters of `Kernel` and `Discriminator` are a free-form but json serializable dictionary (i.e. no numpy, no parameter, etc...). However I've never seen actual spec of these objects. @taalexander do you have any information?